x/evm/keeper/chainKeeper.go

```diff
@@ -644,42 +675,88 @@ func (k chainKeeper) CreateNewBatchToSign(ctx sdk.Context) ([]byte, error) {
 	}
 
 	k.setCommandBatchMetadata(ctx, batchedCommands)
-	k.getStore(ctx, k.chain).SetRaw(unsignedBatchIDKey, batchedCommands.ID)
+	k.setUnsignedBatchID(ctx, batchedCommands.ID)
 
 	return batchedCommands.ID, nil
 }
 
+func (k chainKeeper) setUnsignedBatchID(ctx sdk.Context, id []byte) {
+	k.getStore(ctx, k.chainLowerKey).SetRaw(unsignedBatchIDKey, id)
+}
+
 // returns the queue of commands
-func (k chainKeeper) getCommandQueue(ctx sdk.Context) utils.KVQueue {
-	return utils.NewBlockHeightKVQueue(commandQueueName, k.getStore(ctx, k.chain), ctx.BlockHeight(), k.Logger(ctx))
+func (k chainKeeper) getCommandQueue(ctx sdk.Context) utils.BlockHeightKVQueue {
+	return utils.NewBlockHeightKVQueue(commandQueueName, k.getStore(ctx, k.chainLowerKey), ctx.BlockHeight(), k.Logger(ctx))
 }
 
-func (k chainKeeper) setTokenMetadata(ctx sdk.Context, asset string, meta types.ERC20TokenMetadata) {
+func (k chainKeeper) serializeCommandQueue(ctx sdk.Context) map[string]types.Command {
+	iter := k.getStore(ctx, k.chainLowerKey).Iterator(utils.KeyFromStr(commandQueueName))
+	defer utils.CloseLogError(iter, k.Logger(ctx))
+
+	commands := make(map[string]types.Command)
+	for ; iter.Valid(); iter.Next() {
+		var command types.Command
+		iter.UnmarshalValue(&command)
+		key := string(iter.Key())
+		key = strings.TrimPrefix(key, commandQueueName)
+		key = strings.TrimPrefix(key, "_")
+		commands[key] = command
+	}
+
+	return commands
+}
+
+func (k chainKeeper) setCommandQueue(ctx sdk.Context, queueState map[string]types.Command) {
+	state := make(map[string]codec.ProtoMarshaler, len(queueState))
+	for key, value := range queueState {
+		// need to create a new variable inside the loop because the state map takes its reference,
+		// otherwise all entries would refer to a single &value pointer
```

Comment:
i don't quite follow, but I trust you :P

Reply:
ya of course golang 😞

Reply:
`key` and `value` are reused in every iteration of the loop, so if I use `&value` inside the loop, each map entry will take the same reference pointer to the `value` variable, so in the end they will all have the content of the last `value` of the iteration